tests/optional_kernel_features/kernel_features_device_has_exceptions.cpp

```diff
@@ -0,0 +1,329 @@
+
+      run_functor<FunctorT>(is_exception_expected, expected_errc, queue);
+    }
+
+    {
+      RUN_SUBMISSION_CALL(is_exception_expected, expected_errc, queue,
+                          [[sycl::device_has(FeatureAspectT)]],
+                          dummy_function_non_decorated());
+    }
+  }
+
+  SECTION(
+      "Kernel that calls a function which is decorated with the feature's "
+      "corresponding attribute [[sycl::device_has()]]. Neither the kernel "
+      "nor the function use the feature.") {
+    {
+      const auto lambda_no_arg = []() {
+        dummy_function_decorated<FeatureAspectT>();
+      };
+      const auto lambda_item_arg = [](sycl::item<1>) {
+        dummy_function_decorated<FeatureAspectT>();
+      };
+      const auto lambda_group_arg = [](sycl::group<1>) {
+        dummy_function_decorated<FeatureAspectT>();
+      };
+
+      run_separate_lambda(is_exception_expected, expected_errc, queue,
+                          lambda_no_arg, lambda_item_arg, lambda_group_arg);
+    }
+
+    {
+      using FunctorT = non_decorated_call_decorated_dummy<FeatureAspectT>;
+
+      run_functor<FunctorT>(is_exception_expected, expected_errc, queue);
+    }
+
+    {
+      RUN_SUBMISSION_CALL(is_exception_expected, expected_errc, queue,
+                          NO_ATTRIBUTE,
+                          dummy_function_decorated<FeatureAspectT>());
+    }
+  }
+
+  SECTION(
+      "Kernel that calls a function which is decorated with the feature's "
+      "corresponding attribute [[sycl::device_has()]]. Function uses the "
+      "feature and kernel doesn't.") {
+    {
+      const auto lambda_no_arg = []() {
+        use_feature_function_decorated<FeatureTypeT, FeatureAspectT>();
+      };
+      const auto lambda_item_arg = [](sycl::item<1>) {
+        use_feature_function_decorated<FeatureTypeT, FeatureAspectT>();
+      };
+      const auto lambda_group_arg = [](sycl::group<1>) {
+        use_feature_function_decorated<FeatureTypeT, FeatureAspectT>();
+      };
+
+      run_separate_lambda(is_exception_expected, expected_errc, queue,
+                          lambda_no_arg, lambda_item_arg, lambda_group_arg);
+    }
+
+    {
+      using FunctorT =
+          non_decorated_call_decorated_function<FeatureTypeT, FeatureAspectT>;
+
+      run_functor<FunctorT>(is_exception_expected, expected_errc, queue);
+    }
+
+    {
+      RUN_SUBMISSION_CALL(
+          is_exception_expected, expected_errc, queue, NO_ATTRIBUTE,
+          use_feature_function_decorated<FeatureTypeT, FeatureAspectT>());
+    }
+  }
+
+  SECTION(
+      "Kernel with tested feature but with attribute [[sycl::device_has()]] "
+      "for another feature.") {
+    constexpr sycl::aspect AnotherFeatureAspect =
+        get_another_aspect<FeatureAspectT>();
+    bool other_feature_exception_expect = true;
+    if (queue.device_has(AnotherFeatureAspect) &&
+        queue.device_has(FeatureAspectT)) {
```

Comment:
@andreyromanof, is this a DPC++ extension?

Reply:
No. As far as I know, it is not an extension.

Reply:
I don't see such member function in the specification, so it seems to be illegal. Rigth?

Reply:
You're right. There is only `has(aspect)` member function of `device` class. Looks like I have to change all `queue.device_has(..)` to `queue.get_device().has(..)`